### PR TITLE
[RISCV][VLOPT] Add support for integer multiply-add instructions

### PR DESCRIPTION
**Pull request by michaelmaitland**
This adds support for these instructions.

### CODE REVIEW

**Comment by topperc** [comment]
> I think the VL on the using add instruction can be optimized further, once we add support for optimizing non-vlmax.

I'm not sure what this means since there are no add instructions in the tests.

**Comment by michaelmaitland** [comment]
> > I think the VL on the using add instruction can be optimized further, once we add support for optimizing non-vlmax.
> 
> I'm not sure what this means since there are no add instructions in the tests.

I've removed it from the description. Sorry.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp [hunk 1]
```diff
@@ -550,7 +550,14 @@ static bool isSupportedInstr(const MachineInstr &MI) {
   // 11.12. Vector Widening Integer Multiply Instructions
   // FIXME: Add support for 11.12 instructions
   // 11.13. Vector Single-Width Integer Multiply-Add Instructions
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:552]: One concern that is not related to the code: the [RVV repo](https://github.com/riscv/riscv-v-spec) has been archived, and all the content are put in a section of [riscv-isa-manual](https://github.com/riscv/riscv-isa-manual) now, so these section markers are kind of misleading.

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:552]: Good observation, thanks. I have updated the patch.

File: llvm/test/CodeGen/RISCV/rvv/vl-opt-instrs.ll [hunk 2]
```diff
@@ -1248,6 +1248,203 @@ define <vscale x 4 x i64> @vwmulu_vx(<vscale x 4 x i32> %a, i32 %b, iXLen %vl) {
...
+;
+; VLOPT-LABEL: vmacc_vv:
+; VLOPT:       # %bb.0:
+; VLOPT-NEXT:    vsetvli zero, a0, e32, m2, tu, ma
+; VLOPT-NEXT:    vmacc.vv v8, v8, v10
+; VLOPT-NEXT:    vsetvli zero, zero, e32, m2, ta, ma
+; VLOPT-NEXT:    vmul.vv v8, v8, v10
+; VLOPT-NEXT:    ret
+  %1 = call <vscale x 4 x i32> @llvm.riscv.vmacc.nxv4i32.nxv4i32(<vscale x 4 x i32> %a, <vscale x 4 x i32> %a, <vscale x 4 x i32> %b, iXLen -1, iXLen 0)
+  %2 = call <vscale x 4 x i32> @llvm.riscv.vmul.nxv4i32.nxv4i32(<vscale x 4 x i32> poison, <vscale x 4 x i32> %1, <vscale x 4 x i32> %b, iXLen %vl)
```
Comment (lukel97) [inline llvm/test/CodeGen/RISCV/rvv/vl-opt-instrs.ll:1268]: Nit, use vadd for the user instead of vmul?

File: llvm/test/CodeGen/RISCV/rvv/vl-opt-instrs.ll [hunk 3]
```diff
@@ -1248,6 +1248,203 @@ define <vscale x 4 x i64> @vwmulu_vx(<vscale x 4 x i32> %a, i32 %b, iXLen %vl) {
...
+;
+; VLOPT-LABEL: vmacc_vx:
+; VLOPT:       # %bb.0:
+; VLOPT-NEXT:    vmv2r.v v10, v8
+; VLOPT-NEXT:    vsetvli zero, a1, e32, m2, tu, ma
+; VLOPT-NEXT:    vmacc.vx v10, a0, v8
+; VLOPT-NEXT:    vsetvli zero, zero, e32, m2, ta, ma
+; VLOPT-NEXT:    vmul.vv v8, v10, v8
+; VLOPT-NEXT:    ret
+  %1 = call <vscale x 4 x i32> @llvm.riscv.vmacc.nxv4i32.nxv4i32(<vscale x 4 x i32> %a, i32 %b, <vscale x 4 x i32> %a, iXLen -1, iXLen 0)
```
Comment (lukel97) [inline llvm/test/CodeGen/RISCV/rvv/vl-opt-instrs.ll:1311]: Nit, should this be
```suggestion
  %1 = call <vscale x 4 x i32> @llvm.riscv.vmacc.nxv4i32.i32(<vscale x 4 x i32> %a, i32 %b, <vscale x 4 x i32> %a, iXLen -1, iXLen 0)
```

**Review by lukel97 — APPROVED** [review]

File: llvm/test/CodeGen/RISCV/rvv/vl-opt-instrs.ll (hunk 3 above)
Comment (topperc) [reply llvm/test/CodeGen/RISCV/rvv/vl-opt-instrs.ll:1311]: Shouldn't it be iXLen not i32?

Comment (lukel97) [reply llvm/test/CodeGen/RISCV/rvv/vl-opt-instrs.ll:1311]: Isn't the second type argument for the scalar operand which can be an illegal integer type? E.g. from llvm/test/CodeGen/RISCV/rvv/vmacc.ll:

```llvm
declare <vscale x 4 x i8> @llvm.riscv.vmacc.mask.nxv4i8.i8(
  <vscale x 4 x i8>,
  i8,
  <vscale x 4 x i8>,
  <vscale x 4 x i1>,
  iXLen, iXLen);
```

Comment (topperc) [reply llvm/test/CodeGen/RISCV/rvv/vl-opt-instrs.ll:1311]: oh you're right. I forget that we keep leaving off the VL operand type. There are really 3 overloaded types.

**Comment by topperc** [comment]
Should we have tests where vmacc is the using instruction so we test the operand info for the source operands?

**Comment by topperc** [comment]
> Should we have tests where vmacc is the using instruction so we test the operand info for the source operands?

Oh I guess we have one test `vmacc_vv_use`. But that only tests one of the source operands.

**Comment by topperc** [comment]
> > Should we have tests where vmacc is the using instruction so we test the operand info for the source operands?
> 
> The vop_vv* tests in vl-opt-op-info.mir should also test the same getOperandInfo cases

Are you saying the tests already exist? Or are you asking Michael to add more tests to that file?

**Comment by lukel97** [comment]
> > > Should we have tests where vmacc is the using instruction so we test the operand info for the source operands?
> > 
> > 
> > The vop_vv* tests in vl-opt-op-info.mir should also test the same getOperandInfo cases
> 
> Are you saying the tests already exist? Or are you asking Michael to add more tests to that file?

They already exist

**Review by topperc — APPROVED** [review]
LGTM

**Comment by lukel97** [comment]
I think vl-opt-instrs.ll needs updated, the pre-merge CI seems to be failing
